### ISSUE TITLE
Implicit `this` parameter bypasses assignability checks with derived methods

### ISSUE THREAD

**Issue by 64qq** [microsoft/TypeScript#62055]
### 🔎 Search Terms

contextually typed this, derive this parameter type, implicit this parameter

### 🕗 Version & Regression Information

This is the behavior in every version I tried on the playground

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.8.3#code/MYGwhgzhAEBCkFNoG8CwAoaXoFsEBcALAewBMAKIgSwgC5oJ8AnKgOwHMBKFaAXw37oMoSDACSOAA4gqwKvgAqhGtAQAPfAlakY8CEjSZseImXLdD2K9AD0N6AFE1khME2loYVqqZNiTaEIEJgQMa2xqCAA6L2IiYIBZAhIKTgBuMOxBTKxY+KYk01SeQWz0cgAzAFdWNypib3I2Ri9gBHo9BAsc6Gb8VoQokxSo4DAQEHIAIkBeDcBBPan0nrtoAFVasCr2QnxoBQBPFwdff3pImNY4oILksl6YS92waGra-HrWAU5yVgQAd2gEmksnkSho5iWQnQKycwLk+BA+08rEeYHeHGgf3khGgAHJIrjMf4ANYwSC9ViabQIUjCcBQRzOGTwsEwdRUnRwRAoHrDMyRM7KCDdIzhc55a6FFLmDKivgYHoSxK3YrIeXoMqVGp1BrQJqsFq1dpc-Qiqx9AZDFWjcaTWaAIT3FrLeN9fgDYczQUKIbKMEA

### 💻 Code

```ts
class Base {
    method(this: string) { }
}

class ImplicitThis extends Base {
    method() {
        // Expected an error here
        this.anotherMethod();
    }

    anotherMethod() { }
}

(function (instance: Base) {
    instance.method.call("🙁");
    // Uncaught TypeError: this.anotherMethod is not a function
})(new ImplicitThis());
```

### 🙁 Actual behavior

It is permitted to use class methods with implicit `this` parameters, even when the `this` type is not compatible with that of the derived methods.

### 🙂 Expected behavior

Explicitly annotating with `this` works as intended.
```ts
class ExplicitThis extends Base {
    method(this: this) {
//  ^^^^^^ Property 'method' in type 'ExplicitThis' is not assignable to the same property in base type 'Base'...
        this.anotherMethod();
    }

    anotherMethod() { }
}

(function (instance: Base) {
    instance.method.call("🙂");
})(new ExplicitThis());
// ^^^^^^^^^^^^^^^^^^ Argument of type 'ExplicitThis' is not assignable to parameter of type 'Base'...
```

TypeScript should also report such errors for implicit `this` parameters, or infer the `this` type from the derived methods.

**Comment by RyanCavanaugh** [comment]
Implicitly adding `this: this` to all class methods had catastrophic perf implications, so we don't do this. The best workaround is set up a lint rule to enforce all methods to have a `this` parameter.
